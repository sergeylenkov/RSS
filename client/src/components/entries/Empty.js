import React from 'react';

import styles from './Empty.module.css';

export class EmptyList extends React.Component {
    render() {
        return (
            <div className={styles.container}>
                <div className={styles.header}>Нет новых постов.</div>
                <div className={styles.buttons}>
                    Посмотреть
                    <button className={styles.button} onClick={this.props.onShowLast}>прошлые</button>
                    <button className={styles.button} onClick={this.props.onShowFavorites}>избранное</button>
                    <button className={styles.button} onClick={this.props.onShowRead}>прочитанное</button>
                </div>
            </div>
        );
    }
}