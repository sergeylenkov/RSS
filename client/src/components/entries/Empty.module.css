.container {
    max-width: 800px;
    min-width: 800px;
    margin: 40px auto 0 auto;
	padding: 0;
    display: flex;
    flex-direction: column;
}

.header {
    margin: 0 auto 20px 0;
    font-size: 18px;    
}

.buttons {
    margin: 0 auto 0 0;
    font-size: 14px;
    line-height: 18px;
}

.button {
    margin: 0 5px;
    font-size: 14px;
    line-height: 18px;
    color: #0060a0;
    border-bottom: 1px dashed rgba(0, 96, 160, 0.4);
    cursor: pointer;
}