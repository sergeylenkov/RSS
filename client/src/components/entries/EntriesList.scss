.entries-list {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.grid .entries-list {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-column-gap: 20px;
}

@media (min-width: 1100px) {
  .grid .entries-list {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}
