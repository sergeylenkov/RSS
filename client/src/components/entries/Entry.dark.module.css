.container {
    background-color: #1d1d1d;
}

.title a {
	color: #bfbfbf;
}

.description {
    color: #bfbfbf;
}

.description font {
	color: #bfbfbf !important;
}

.expandButton {
    background-color: rgba(29, 29, 29, 0.8);
}