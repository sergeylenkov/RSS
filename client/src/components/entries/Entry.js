import React from 'react';

import styles from './Entry.module.css';

export class Entry extends React.Component {
    render() {
        const description = this.clearSelfLinks(this.props.description, this.props.link);
        
        return (
            <div className={styles.container}>
                <div className={styles.title}><a href={this.props.link}>{this.props.title}</a></div>
                <div className={styles.description} dangerouslySetInnerHTML={{ __html: description }}></div>
            </div>
        );
    }

    clearSelfLinks(description, link) {
        let el = document.createElement('div');       
        el.innerHTML = description;
        
        let items = Array.from(el.getElementsByTagName('a'));

        let links = items.filter((item) => {
            let href = item.getAttribute('href');
    
            if (href && href.indexOf(link) !== -1) {
                return true;
            }
    
            return false;
        });
    
        for (let i = 0; i < links.length; i++) {
            links[i].remove();
        }
    
        return el.innerHTML;
    }    
}