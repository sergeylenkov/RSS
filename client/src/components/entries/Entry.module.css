.container {
	position: relative;
	width: 100%;
    margin: 0 0 20px 0;
    padding: 20px;
    box-sizing: border-box;
    flex-shrink: 0;    
    background: #ffffff;
}

.container.read {
	opacity: 0.8;
}

.title {
	font-size: 18px;
	font-weight: bold;
	line-height: 22px;
	margin: 0 0 15px 0;
}

.title a {
	color: #000000;
	text-decoration: none;
}

.description {
	position: relative;
	font-size: 14px;
	line-height: 18px;
	overflow: hidden;
	margin: 20px 0 0 0;
}

.description img {
	max-width: 100%;
	padding: 0px;
	height: auto;
    margin: 0 auto 10px auto;
}

.description font {
	color: #000000 !important;
}

.description p {
	margin: 0 0 10px 0;
}

.description code {
	font-family: inherit;
}

.description blockquote {
	font-style: italic;
    color: #767676;
	margin: 0;
}

.description h1,
.description h2 {
	font-size: 14px;
    margin: 0;
}

.description ol {
	margin: 0;
    padding: 0 0 0 20px;
}

.description a {
	color: #4076cc;
}

.feed {	
	margin: 10px 0;	
	display: flex;
}

.feedIcon {
	width: 14px;
	height: 14px;
	margin: auto 0;
	background-image: "";
    background-repeat: no-repeat;
    background-size: cover;
    background-color: #ebebeb;
}

.feedTitle {
	font-size: 12px;
	margin: auto 0 auto 10px;
	color: #909090;
}

.collapsed .description {
	max-height: 800px;
}

.expandButton {
	position: absolute;	
	left: 0;
	right: 0;
	bottom: 20px;
	height: 60px;
	width: 100%;
	background: linear-gradient(to bottom, rgba(255, 255, 255, 0.1), rgba(255, 255, 255, 0.8));
	cursor: pointer;
}