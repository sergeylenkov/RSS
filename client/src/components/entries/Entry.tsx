import React, {createRef} from 'react';
import { connect } from 'react-redux';
import { FavoriteSelectedIcon, FavoriteIcon, ReadIcon } from '../Icons';
import { Entry } from "../../data/data";

import lightStyles from './Entry.module.css';
import darkStyles from './Entry.dark.module.css';

interface MapStateToProps {
  isDarkTheme: boolean;
}

interface EntryProps extends MapStateToProps {
  entry: Entry;
  isCollapseLong: boolean;
  onView: (id: number) => void;
  onSetRead: (id: number, isRead: boolean) => void;
  onSetFavorite: (id: number, isRead: boolean) => void;
}

interface EntryState {
  isExpanded: boolean;
}

class EntryItem extends React.Component<EntryProps, EntryState> {
  public state: EntryState = {
    isExpanded: false
  };

  private itemElement: HTMLDivElement | null = null;
  private titleElement: HTMLDivElement | null = null;

  private itemRefCallback = (element: HTMLDivElement) => {
    if (element) this.itemElement = element;
  };

  private titleRefCallback = (element: HTMLDivElement) => {
    if (element) {
      this.titleElement = element;
      this.titleElement.addEventListener('mouseup', this.onRead);
    }
  };

  constructor(props: Readonly<EntryProps>) {
    super(props);

    this.handleScroll = this.handleScroll.bind(this);
    this.handleMove = this.handleMove.bind(this);
  }

  public componentDidMount() {
    if (!this.props.entry.isViewed) {
      window.addEventListener('scroll', this.handleScroll);
      window.addEventListener('mousemove', this.handleMove);
    }
  }

  public componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
    window.removeEventListener('mousemove', this.handleMove);
  }

  private removeSelfLinks(description: string, link: string) {
    let el = document.createElement('div');
    el.innerHTML = description;

    let items = Array.from(el.getElementsByTagName('a'));

    let links = items.filter((item) => {
      let href = item.getAttribute('href');

      return !!(href && href.indexOf(link) !== -1);
    });

    for (let i = 0; i < links.length; i++) {
      links[i].remove();
    }

    return el.innerHTML;
  }

  static isLong(description: string) {
    let el = document.createElement('div');
    el.innerHTML = description;

    if (el.getElementsByTagName('img').length > 3) {
      return true;
    }

    return el.innerText.length > 1500;
  }

  private handleScroll() {
    const { entry, onView } = this.props;

    if (!entry.isViewed) {
      const rect = this.itemElement!.getBoundingClientRect();
      const height = window.innerHeight / 2;

      let isViewed = false;
      let scrollEnd = false;

      if (window.innerHeight + window.scrollY >= document.body.scrollHeight) {
        scrollEnd = true;
      }

      if (scrollEnd && rect.top > 0) {
        isViewed = true;
      } else if (rect.top < height) {
        isViewed = true;
      }

      if (isViewed) {
        onView(entry.id);
      }
    }
  }

  private handleMove() {
    if (!this.props.entry.isViewed) {
      const rect = this.itemElement!.getBoundingClientRect();
      const height = rect.top + rect.height;

      if (height < window.innerHeight) {
        this.props.onView(this.props.entry.id);
      }

      window.removeEventListener('mousemove', this.handleMove);
    }
  }

  private onExpand = () => {
    this.setState({
      isExpanded: true
    });

    this.onRead();
  };

  private onRead = () => {
    const { onSetRead, entry } = this.props;

    if (!entry.isRead) {
      onSetRead(entry.id, true);
    }
  };

  private onUnread = () => {
    const { onSetRead, entry } = this.props;
    onSetRead(entry.id, false);
  };

  private onSetFavorite = () => {
    const { onSetFavorite, entry } = this.props;
    onSetFavorite(entry.id, !entry.isFavorite);
  };

  public render() {
    const { isDarkTheme, entry, isCollapseLong } = this.props;
    const { isExpanded } = this.state;

    let styles = lightStyles;

    if (isDarkTheme) {
      styles = {...lightStyles, ...darkStyles};
    }

    const description = this.removeSelfLinks(entry.description, entry.link);

    let readIcon = null;

    if (entry.isRead) {
      readIcon = <div className={styles.infoItem} onClick={this.onUnread}><div className={styles.infoIcon}><ReadIcon /></div></div>;
    }

    let className = styles.container;
    let expandButton = null;

    if (isCollapseLong && !isExpanded && EntryItem.isLong(description)) {
      className += ` ${styles.collapsed}`;
      expandButton = <button className={styles.expandButton} onClick={this.onExpand} />
    }

    return (
      <div className={className} ref={this.itemRefCallback}>
        <div
          className={styles.title}
          ref={this.titleRefCallback}>
          <a href={entry.link}>{entry.title}</a>
        </div>
        <div className={styles.feed}>
          <div className={styles.feedIcon} style={{ backgroundImage: `url(${entry.feed.icon})` }} />
          <div className={styles.feedTitle}>{entry.feed.title}</div>
        </div>
        <div className={styles.description} dangerouslySetInnerHTML={{ __html: description }} />
        {expandButton}
        <div className={styles.info}>
          <div className={styles.infoItem}>
            <div className={styles.infoIcon} onClick={this.onSetFavorite}>
              {entry.isFavorite ? <FavoriteSelectedIcon /> : <FavoriteIcon /> }
            </div>
          </div>
          {readIcon}
        </div>
      </div>
    );
  }
}

/* Redux */

const mapStateToProps = (state: MapStateToProps) => {
  return {
    isDarkTheme: state.isDarkTheme
  };
};

export default connect(mapStateToProps)(EntryItem);
