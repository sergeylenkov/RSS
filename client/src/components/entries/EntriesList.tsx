import React, { useEffect, useRef } from 'react';
import { useLocation } from 'react-router-dom';
import { entriesUpdated, updateEntriesCount, updateFavorite, updateRead, updateUnviewedCount, updateViewed } from '../../store/actions';
import { useDispatch, useSelector } from 'react-redux';
import Data, { Entry } from '../../data';
import { debounce } from '../../utils';
import { State } from '../../store/reducers';
import EntryItem from './EntryItem';

import './EntriesList.scss';

function EntriesList(): JSX.Element {
  const { pathname } = useLocation();
  const isInitialized = useSelector<State, boolean>(state => state.isInitialized);
  const isCollapseLong = useSelector<State, boolean>(state => state.isCollapseLong);
  const entries = useSelector<State, Entry[]>(state => state.entries);
  const dispatch = useDispatch();
  const entryItems = useRef<Map<number, HTMLDivElement>>(new Map()).current;
  const viewedIds = useRef<Set<number>>(new Set()).current;
  const observer = useRef<IntersectionObserver>();

  const OnViewDebonce = debounce(() => {
    Data.setViewed([...viewedIds]).then(() => {
      dispatch(updateViewed([...viewedIds]));
      dispatch(updateUnviewedCount());

      viewedIds.clear();
    });
  }, 1000, false);

  const onView = (id: number) => {
    viewedIds.add(id);
    OnViewDebonce();
  };

  const onSetRead = (id: number, isRead: boolean) => {
    Data.setRead(id, isRead).then(() => {
      dispatch(updateRead(id, isRead));
    });
  };

  const onSetFavorite = (id: number, isFavorite: boolean) => {
    Data.setFavorite(id, isFavorite).then(() => {
      dispatch(updateFavorite(id, isFavorite));
    });
  };

  const getEntries = (path: string) => {
    if (!isInitialized) {
      return;
    }

    if (path === '/') {
      Data.getUnviewed().then((response: Entry[]) => {
        dispatch(entriesUpdated(response));
        dispatch(updateUnviewedCount());
      });
    }

    if (path === '/all') {
      Data.allEntries().then((response: Entry[]) => {
        dispatch(entriesUpdated(response));
        dispatch(updateEntriesCount());
      });
    }

    if (path === '/read') {
      Data.readEntries().then((response: Entry[]) => {
        dispatch(entriesUpdated(response));
        dispatch(updateEntriesCount());
      });
    }

    if (path === '/favorites') {
      Data.getFavorites().then((response: Entry[]) => {
        dispatch( entriesUpdated(response));
        dispatch(updateEntriesCount());
      });
    }
  }

  useEffect(() => {
    getEntries(pathname);
  }, [pathname]);

  useEffect(() => {
    observer.current = new IntersectionObserver((entries) => {
      entries.forEach(el => {
        if (el.isIntersecting) {
          for (const [key, value] of entryItems) {
            if (value === el.target) {
              onView(key);
            }
          }
        }
      });
    }, { threshold: 0.5 });

    return () => {
      if (observer.current) {
        observer.current.disconnect();
      }
    }
  }, [])

  return (
    <div className='entries-list'>
      {
        entries.map((entry: Entry) => {
          return (
            <EntryItem
              ref={(ref) => {
                if (ref && observer.current && !entry.isViewed) {
                  entryItems.set(entry.id, ref);
                  observer.current.observe(ref);
                }
              }}
              key={entry.id}
              entry={entry}
              isCollapseLong={isCollapseLong}
              onSetRead={onSetRead}
              onSetFavorite={onSetFavorite} />
          )
        })
      }
    </div>
  );
}

export default EntriesList;
