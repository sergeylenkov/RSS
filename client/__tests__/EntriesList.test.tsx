import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../src/store';
import { EntriesList } from '../src/components/entries/EntriesList';

test('EntriesList', () => {
  const render = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <EntriesList />
      </BrowserRouter>
    </Provider>
  );
  expect(render).toMatchSnapshot();
});
