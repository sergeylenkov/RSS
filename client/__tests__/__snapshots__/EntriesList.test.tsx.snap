// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu correct rendering 1`] = `
<div
  className="entries-list"
/>
`;
